devito/finite_differences/finite_difference.py

```diff
@@ -0,0 +1,401 @@
+    :param order: Discretisation order of the stencil to create.
+    """
+    first = generic_derivative(expr, deriv_order=deriv_order[0],
+                               fd_order=fd_order[0], dim=dims[0])
+    return generic_derivative(first, deriv_order=deriv_order[1],
+                              fd_order=fd_order[1], dim=dims[1])
+
+
+def staggered_diff(expr, deriv_order, dim, fd_order, stagger=centered):
+    """
+    Utility function to generate staggered derivatives
+    :param f: function objects, eg. `f(x, y)` or `g(t, x, y, z)`.
+    :param dims: symbol defining the dimension wrt. which
+       to differentiate, eg. `x`, `y`, `z` or `t`.
+    :param order: Order of the coefficient discretization and thus
+                  the width of the resulting stencil expression.
+    :param stagger: Shift for the FD, `left`, `right` or `centered`
+    :param theta: Dip (or polar) angle for rotated FD
+    :param phi: Azimuth angle for rotated FD
+    """
+
+    if stagger == left:
+        off = -.5
+    elif stagger == right:
+        off = .5
+    else:
+        off = 0
+    diff = dim.spacing
+    idx = list(set([(dim + int(i+.5+off)*diff)
+                    for i in range(-int(fd_order / 2), int(fd_order / 2))]))
+    if int(fd_order / 2) == 1:
+        idx = [dim - diff, dim]
+    c = finite_diff_weights(deriv_order, idx, dim + off*dim.spacing)[-1][-1]
+    deriv = 0
+    for i in range(0, len(idx)):
+            deriv += expr.subs({dim: idx[i]}) * c[i]
+
+    return deriv.evalf(_PRECISION)
+
+
+def staggered_cross_diff(expr, dims, deriv_order, fd_order, stagger):
+    """
+    Create a generic cross derivative for a given staggered function.
+
+    :param function: The symbol representing a function.
+    :param dims: Dimensions for which to take the derivative.
+    :param order: Discretisation order of the stencil to create.
+    """
+    first = staggered_diff(expr, deriv_order=deriv_order[0],
+                           fd_order=fd_order[0], dim=dims[0], stagger=stagger[0])
+    return staggered_diff(first, deriv_order=deriv_order[1],
+                          fd_order=fd_order[1], dim=dims[1], stagger=stagger[1])
+
+
+def generate_fd_functions(function):
+    """
+    Create all possible FD functions from the indices,
+    space_order (and time_order if time dependant)
+    """
+    dimensions = function.indices
+    space_fd_order = function.space_order
+    time_fd_order = function.time_order if function.is_TimeFunction else 0
+    is_staggered = any(s is not None for s in function.staggered)
+    deriv_function = staggered_diff if is_staggered else generic_derivative
+    c_deriv_function = staggered_cross_diff if is_staggered else generic_cross_derivative
+
+    side = dict()
+    for (d, s) in zip(dimensions, function.staggered):
+        if s == 0:
+            side[d] = left
+        elif s == 1:
+            side[d] = right
+        else:
+            side[d] = centered
+
+    derivatives = ()
```

Comment:
list

Reply:
edit: lists are mutable, tuples are immutable.
a list is the right fit here, because you're populating it

Reply:
oki